### PR TITLE
use bindgen to make linux hidraw access portable

### PR DESCRIPTION
**Pull request by estokes**
here is my proposed fix to #96

### CODE REVIEW

**Comment by baloo** [comment]
that would require linux headers on the builder, I'm not sure that's available with firefox build environment. cc @jcjones

**Comment by baloo** [comment]
an alternative would be to embed ppc64 and other headers like:
``` rust
# build.rs#[cfg(feature = "binding-recompile")]
#[cfg(not(feature = "binding-recompile"))]
extern crate bindgen;

#[cfg(not(feature = "binding-recompile"))]
use std::env;
#[cfg(not(feature = "binding-recompile"))]
use std::path::PathBuf;

#[cfg(or(not(target_os = "linux"), not(feature = "binding-recompile")))]
fn main () {}

#[cfg(and(target_os = "linux", feature = "binding-recompile))]
fn main() {
    let bindings = bindgen::Builder::default()
        .header("src/linux/hidwrapper.h")
        .whitelist_var("_HIDIOCGRDESCSIZE")
        .whitelist_var("_HIDIOCGRDESC")
        .generate()
        .expect("Unable to get hidraw bindings");

    let out_path = PathBuf::from("src/linux/");
    // Add $arch in out_path 
    bindings
        .write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write hidraw bindings");
}
```

and commit the output in the git.

**Comment by estokes** [comment]
We could do that, but it seems like it's getting pretty complex, we're going to require a manual step to refresh the headers, and then commit them to the repo. I think I would rather just use the approach used in https://github.com/jmesmon/ioctl/blob/master/ioctl-sys/src/platform/linux.rs I don't think these constants will change often, and it's likely just as easy to copy new ones into the rust code and commit, vs compile with a special feature and commit. This would also let us ditch the dependency on bindgen.

Of course it would be much better if the build box had the kernel headers installed, given the huge scope of firefox I'm surprised they aren't already required. I will wait to rework the patch until we hear an answer about that.

**Comment by estokes** [comment]
Actually, now that I think about it more, taking into account endianness makes your solution the best one, I will work on it.

**Comment by estokes** [comment]
Ok, it's done, please have a look

**Comment by baloo** [comment]
LGTM

File: src/linux/hidraw.rs [hunk 1]
```diff
@@ -19,42 +20,26 @@ pub struct LinuxReportDescriptor {
...
-const SIZESHIFT: u32 = TYPESHIFT + TYPEBITS as u32;
-const DIRSHIFT: u32 = SIZESHIFT + SIZEBITS as u32;
-
-// https://github.com/torvalds/linux/blob/master/include/uapi/linux/hid.h
 const HID_MAX_DESCRIPTOR_SIZE: usize = 4096;
 
-macro_rules! ioctl {
-    ($dir:expr, $name:ident, $ioty:expr, $nr:expr; $ty:ty) => {
-        pub unsafe fn $name(fd: libc::c_int, val: *mut $ty) -> io::Result<libc::c_int> {
-            let size = mem::size_of::<$ty>();
```
Comment (jcjones) [inline src/linux/hidraw.rs]: Clippy points out `std::mem` is now unused, so going to pull that out.

**Review by jcjones — APPROVED** [review]
LGTM to me too-- thank you!

**Comment by glaubitz** [comment]
Could you add any instructions on how to use bindgen here? I tried building authenticator-rs on sparc64 and don't know how to extract the ioctl constants.

**Pull request merged**